readme.md

![Bilby Stampede](http://turnjs.com/pics/small-turnjs-letters.png)

##Get the [turn.js 4th release](https://github.com/Estupendo/turn.js/releases)


### What's new in turn.js 4th release?

- Added option `autoCenter`

- Added option `zoom`

- Added property `animating`

- Added property `zoom`

- Added method `center`

- Added method `destroy`

- Added method `is`

- Added method `zoom`

- Added event `missing`

- Added event `zooming`

- Added class `.even`

- Added class `.fixed`

- Added class `.hard`

- Added class `.odd`

- Added class `.own-size`

- Added class `.sheet`

- Added the ignore attribute

- New turn.html4.js

- New scissors.js

- Changed the class `.turn-page` to `.page`

- Improved the animation frame generator with requestAnimationFrame

- Improved the animation speed for hard pages with CSS3 transitions

- Redesigned the event sequence to listen to only three events

- Fixed issue #79

- Fixed issue #91

- Fixed issue about the event order turning + turned

- Fixed issue about appending pages in wrong locations

Available only on [turnjs.com](http://www.turnjs.com/)

* * *

turn.js 3rd release
=========

### Make a flip book with HTML5

Turn.js is a plugin for jQuery that adds a beautiful transition similar to real pages in a book or magazine. It works in all modern browsers including touch devices.

### What's new?

- New `addPage` for creating pages dynamically.

- New `display` for single and double pages.

- Gradients for non-webkit browsers.

#### Usage

**CSS code:**
```css
#magazine{
	width: 800px;
	height: 400px;
}
#magazine .turn-page{
	background-color:#ccc;
}
```

**HTML code:**
```html
<div id="magazine">
	<div><span class="text">Page 1</span></div>
	<div><span class="text">Page 2</span></div>
	<div><span class="text">Page 3</span></div>
</div>
```

**JavaScript code:**
```javascript
$('#magazine').turn({gradients: true, acceleration: true});
```

#### Requirements

jQuery 1.7 or later

#### Browser support
* Chrome 12, Safari 5, Firefox 10, IE 9

#### License
Released under a non-commercial BSD license

[Full documentation](https://github.com/blasten/turn.js/wiki/Reference)

* * *

[turnjs.com](http://www.turnjs.com/)


changelog.txt
Turn.js
www.turnjs.com

------------------------------------------------------
Turn.js 4.1.0 - 2012/11/65
------------------------------------------------------

+ Added support to the Android's default browser
+ Fixed the hard page effect in IE9
+ Fixed issue #220

------------------------------------------------------
Turn.js 4.0.9 - 2012/08/03
------------------------------------------------------

+ Fixed issue when using the .own-size class
+ Event zoom.change now is preventable

------------------------------------------------------
Turn.js 4.0.8 - 2012/07/16
------------------------------------------------------

+ Added option direction
+ Added method direction
+ Added property direction

------------------------------------------------------
Turn.js 4.0.7 - 2012/07/12
------------------------------------------------------

+ Fixed a issue of the pages method
+ Fixed issue #112
+ corner argument of the start event now returns null when using the methods next, previous and page
+ Fixed issue of certain samples in IE8

------------------------------------------------------
Turn.js 4.0.6 - 2012/07/10
------------------------------------------------------

+ Fixed a issue of the peel method when using the single display
+ Fixed issue of using all the corners in single display mode
+ Added method version
+ Added option turnCorner

------------------------------------------------------
Turn.js 4.0.5 - 2012/06/16
------------------------------------------------------

+ Fixed some visualization issues of the zoom viewport

------------------------------------------------------
Turn.js 4.0.4 - 2012/06/16
------------------------------------------------------

+ Fixed a problem of the method page and event start

------------------------------------------------------
Turn.js 4.0.3 - 2012/06/16
------------------------------------------------------

+ Fixed minor issues of HTML4 version

------------------------------------------------------
Turn.js 4.0.2 - 2012/06/16
------------------------------------------------------

+ New Zoom viewport
+ New magazine sample

------------------------------------------------------
Turn.js 4.0.1 - 2012/06/12
------------------------------------------------------

+ Fixed some issues when calculating z-index of pages
+ Added method peel

------------------------------------------------------
Release 4 Commercial - 2012/06/07
------------------------------------------------------

+ Added option autoCenter
+ Added option zoom
+ Added property animating
+ Added property zoom
+ Added method center
+ Added method destroy
+ Added method is
+ Added method zoom
+ Added event missing
+ Added event zooming
+ Added class .even
+ Added class .fixed
+ Added class .hard
+ Added class .odd
+ Added class .own-size
+ Added class .sheet
+ Added class .sheet
+ Added the ignore attribute
+ New turn.html4.js
+ New scissors.js
+ Changed the class .turn-page to .page
+ Improved the animation frame generator with requestAnimationFrame
+ Improved the animation speed for hard pages with CSS3 transitions
+ Redesigned the event sequence to listen to only three events
+ Fixed issue #79
+ Fixed issue #91
+ Fixed issue about the event order turning
+ turned
+ Fixed issue about appending pages in wrong locations

------------------------------------------------------
Release 3 - 2012/03/01
------------------------------------------------------

+ Added 'range'
+ Added 'addPage'
+ Added 'removePage'
+ Added 'hasPage'
+ Added 'pages'
+ Added 'display'
+ Added 'when' to the initial configuration
+ Added 'pages' to the initial configuration
+ Added 'inclination' to the initial configuration
+ Added 'first' event
+ Added 'last' event
+ Added gradients for non-webkit browsers

------------------------------------------------------
Release 2 - 2012/02/15
------------------------------------------------------

+ Added 'size'.
+ Bug in Chrome 17-18 Beta about losing background-image was fixed.

------------------------------------------------------
Release 1 - 2012/02/05
------------------------------------------------------
+ First alpha release 



samples/steve-jobs/js/steve-jobs.js
/* Steve jobs' book */

function updateDepth(book, newPage) {

	var page = book.turn('page'),
		pages = book.turn('pages'),
		depthWidth = 16*Math.min(1, page*2/pages);

		newPage = newPage || page;

	if (newPage>3)
		$('.sj-book .p2 .depth').css({
			width: depthWidth,
			left: 20 - depthWidth
		});
	else
		$('.sj-book .p2 .depth').css({width: 0});

		depthWidth = 16*Math.min(1, (pages-page)*2/pages);

	if (newPage<pages-3)
		$('.sj-book .p111 .depth').css({
			width: depthWidth,
			right: 20 - depthWidth
		});
	else
		$('.sj-book .p111 .depth').css({width: 0});

}

function loadPage(page) {

	$.ajax({url: 'pages/page' + page + '.html'}).
		done(function(pageHtml) {
			$('.sj-book .p' + page).html(pageHtml.replace('samples/steve-jobs/', ''));
		});

}

function addPage(page, book) {

	var id, pages = book.turn('pages');

	if (!book.turn('hasPage', page)) {

		var element = $('<div />',
			{'class': 'own-size',
				css: {width: 460, height: 582}
			}).
			html('<div class="loader"></div>');

		if (book.turn('addPage', element, page)) {
			loadPage(page);
		}

	}
}

function numberOfViews(book) {

	return book.turn('pages') / 2 + 1;

}

function getViewNumber(book, page) {

	return parseInt((page || book.turn('page'))/2 + 1, 10);

}

function zoomHandle(e) {

	if ($('.sj-book').data().zoomIn)
		zoomOut();
	else if (e.target && $(e.target).hasClass('zoom-this')) {
		zoomThis($(e.target));
	}

}

function zoomThis(pic) {

	var	position, translate,
		tmpContainer = $('<div />', {'class': 'zoom-pic'}),
		transitionEnd = $.cssTransitionEnd(),
		tmpPic = $('<img />'),
		zCenterX = $('#book-zoom').width()/2,
		zCenterY = $('#book-zoom').height()/2,
		bookPos = $('#book-zoom').offset(),
		picPos = {
			left: pic.offset().left - bookPos.left,
			top: pic.offset().top - bookPos.top
		},
		completeTransition = function() {
			$('#book-zoom').unbind(transitionEnd);

			if ($('.sj-book').data().zoomIn) {
				tmpContainer.appendTo($('body'));

				$('body').css({'overflow': 'hidden'});
				
				tmpPic.css({
					margin: position.top + 'px ' + position.left+'px'
				}).
				appendTo(tmpContainer).
				fadeOut(0).
				fadeIn(500);
			}
		};

		$('.sj-book').data().zoomIn = true;

		$('.sj-book').turn('disable', true);

		$(window).resize(zoomOut);
		
		tmpContainer.click(zoomOut);

		tmpPic.load(function() {
			var realWidth = $(this)[0].width,
				realHeight = $(this)[0].height,
				zoomFactor = realWidth/pic.width(),
				picPosition = {
					top:  (picPos.top - zCenterY)*zoomFactor + zCenterY + bookPos.top,
					left: (picPos.left - zCenterX)*zoomFactor + zCenterX + bookPos.left
				};


			position = {
				top: ($(window).height()-realHeight)/2,
				left: ($(window).width()-realWidth)/2
			};

			translate = {
				top: position.top-picPosition.top,
				left: position.left-picPosition.left
			};

			$('.samples .bar').css({visibility: 'hidden'});
			$('#slider-bar').hide();
			
		
			$('#book-zoom').transform(
				'translate('+translate.left+'px, '+translate.top+'px)' +
				'scale('+zoomFactor+', '+zoomFactor+')');

			if (transitionEnd)
				$('#book-zoom').bind(transitionEnd, completeTransition);
			else
				setTimeout(completeTransition, 1000);

		});

		tmpPic.attr('src', pic.attr('src'));

}

function zoomOut() {

	var transitionEnd = $.cssTransitionEnd(),
		completeTransition = function(e) {
			$('#book-zoom').unbind(transitionEnd);
			$('.sj-book').turn('disable', false);
			$('body').css({'overflow': 'auto'});
			moveBar(false);
		};

	$('.sj-book').data().zoomIn = false;

	$(window).unbind('resize', zoomOut);

	moveBar(true);

	$('.zoom-pic').remove();
	$('#book-zoom').transform('scale(1, 1)');
	$('.samples .bar').css({visibility: 'visible'});
	$('#slider-bar').show();

	if (transitionEnd)
		$('#book-zoom').bind(transitionEnd, completeTransition);
	else
		setTimeout(completeTransition, 1000);
}


function moveBar(yes) {
	if (Modernizr && Modernizr.csstransforms) {
		$('#slider .ui-slider-handle').css({zIndex: yes ? -1 : 10000});
	}
}

function setPreview(view) {

	var previewWidth = 115,
		previewHeight = 73,
		previewSrc = 'pages/preview.jpg',
		preview = $(_thumbPreview.children(':first')),
		numPages = (view==1 || view==$('#slider').slider('option', 'max')) ? 1 : 2,
		width = (numPages==1) ? previewWidth/2 : previewWidth;

	_thumbPreview.
		addClass('no-transition').
		css({width: width + 15,
			height: previewHeight + 15,
			top: -previewHeight - 30,
			left: ($($('#slider').children(':first')).width() - width - 15)/2
		});

	preview.css({
		width: width,
		height: previewHeight
	});

	if (preview.css('background-image')==='' ||
		preview.css('background-image')=='none') {

		preview.css({backgroundImage: 'url(' + previewSrc + ')'});

		setTimeout(function(){
			_thumbPreview.removeClass('no-transition');
		}, 0);

	}

	preview.css({backgroundPosition:
		'0px -'+((view-1)*previewHeight)+'px'
	});
}

function isChrome() {

	// Chrome's unsolved bug
	// http://code.google.com/p/chromium/issues/detail?id=128488

	return navigator.userAgent.indexOf('Chrome')!=-1;

}